File: bot.py
import os, tweepy

from datetime import datetime

from comic_dict import comic_dict

PROD = os.environ.get('PROD') == 'true'

if PROD:
    # Production environment, load variables from os.environ
    CONSUMER_KEY = os.environ.get('CONSUMER_KEY')
    CONSUMER_SECRET = os.environ.get('CONSUMER_SECRET')
    ACCESS_SECRET = os.environ.get('ACCESS_SECRET')
    ACCESS_KEY = os.environ.get('ACCESS_KEY')
else:
    # Development environment, load variables from .env file
    from dotenv import load_dotenv
    load_dotenv()
    CONSUMER_KEY = os.getenv('CONSUMER_KEY')
    CONSUMER_SECRET = os.getenv('CONSUMER_SECRET')
    ACCESS_SECRET = os.getenv('ACCESS_SECRET')
    ACCESS_KEY = os.getenv('ACCESS_KEY')

auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)

twitterV1 = tweepy.API(auth)

twitterV2 = tweepy.Client(
    consumer_key=CONSUMER_KEY,
    consumer_secret=CONSUMER_SECRET,
    access_token=ACCESS_KEY, 
    access_token_secret=ACCESS_SECRET
    )


first_day = datetime(2021, 10, 6)


def follow_users():
    targetUser = 'Calvinn_Hobbes'
    # targetId = '1579422614'

    followerIds = twitterV1.get_follower_ids(screen_name=targetUser)
    for id in followerIds[50:60]:
        try:
            twitterV1.create_friendship(user_id=id)
        except:
            print('Unable to follow: ', id)



def get_comic_index():
    return (datetime.now() - first_day).days

def tweet_next_comic():
    comic_index = get_comic_index()
    filename = comic_dict[comic_index]

    print(f"Comic {comic_index}: {filename}")
	
    year = filename[0:4]
    month = filename[4:6]
    day = filename[6:8]

    full_path = f'comics/{year}/{month}/{filename}'

    # if PROD:
    media = twitterV1.media_upload(filename=full_path, chunked=True)
    # else:
    #     print('Upload Media from: ' + full_path)

    text = f'{int(month)}/{int(day)}/{int(year)}'

    # if PROD:
    twitterV1.update_status(status=text, media_ids=[media.media_id])
    # else:
    #     print('Send Tweet w/ text: ' + text)

File: src/bot.ts
import { addDays } from 'date-fns';
import { config } from 'dotenv';
import path from 'path';
import { TwitterApi } from 'twitter-api-v2';
config();

const USERNAME = 'CalvinB0t'

// Twitter API credentials
const consumerKey = process.env.CONSUMER_KEY!;
const consumerSecret = process.env.CONSUMER_SECRET!;
const accessToken = process.env.ACCESS_KEY!;
const accessSecret = process.env.ACCESS_SECRET!;

// Initialize Twitter client
const twitterClient = new TwitterApi({
    appKey: consumerKey,
    appSecret: consumerSecret,
    accessToken: accessToken,
    accessSecret: accessSecret,
});

const FIRST_TWEET_DATE = new Date(2021, 9, 6);

async function getLastTweetDate(): Promise<Date> {
    try {
        const tweets = await twitterClient.v2.userTimeline(process.env.BOT_USER_ID!, {
            max_results: 5,
            "tweet.fields": ["created_at"],
        });

        console.log('tweets', tweets)

        for (const tweet of tweets.data.data) {
            // Try to parse date from tweet text (format: M/D/YYYY)
            const match = tweet.text.match(/(\d{1,2})\/(\d{1,2})\/(\d{4})/);
            if (match) {
                const [_, month, day, year] = match;
                return new Date(parseInt(year), parseInt(month) - 1, parseInt(day));
            }
        }
    } catch (error) {
        console.error('Error getting last tweet date:', error);
    }


    throw new Error('Error getting last tweet date');
}

export const followUsers = async (): Promise<void> => {
    try {
        const targetUser = 'Calvinn_Hobbes';
        const followers = await twitterClient.v2.followers(targetUser);

        for (const follower of followers.data.slice(50, 60)) {
            try {
                await twitterClient.v1.createFriendship({ user_id: follower.id });
                console.log(`Followed user: ${follower.username}`);
            } catch (error) {
                console.error(`Unable to follow user ${follower.username}:`, error);
            }
        }
    } catch (error) {
        console.error('Error in followUsers:', error);
    }
};

export const tweetNextComic = async (): Promise<void> => {
    try {
        const lastTweetDate = await getLastTweetDate();
        const nextTweetDate = addDays(lastTweetDate, 1);

        const year = nextTweetDate.getFullYear();
        const month = nextTweetDate.getMonth() + 1;
        const day = nextTweetDate.getDate();

        const filename = `${year}${month}${day}.gif`;

        // Upload media
        const mediaId = await twitterClient.v1.uploadMedia(filename);

        // Create tweet
        const text = `${month}/${day}/${year}`;
        await twitterClient.v2.tweet({
            text,
            media: { media_ids: [mediaId] }
        });

        console.log('Successfully tweeted comic:', text);
    } catch (error) {
        console.error('Error in tweetNextComic:', error);
    }
}; 